### PR TITLE
fix volume versioning

### PR DESCRIPTION
**Pull request by yasmin-abdullah**
This change relates to 2 stories:

1- [Missing MatchExpressions in Deploy Manifest](https://www.pivotaltracker.com/story/show/177144386). Since this is really small it wasn't worth a deployment in of itself, so I included it here in this PR. 

[This](https://github.com/homedepot/go-clouddriver/blob/bug-version-volumes/pkg/http/core/kubernetes/deploy.go#L158) is the reference in the code for this change

2- [Fix versioning volumes bug](https://www.pivotaltracker.com/story/show/177304155): the previous version of the code would only bind artifacts that are present in the list of the required artifacts for the stage and was breaking when we use one Deploy manifest stage for defining the artifact and the deployment that is binding that artifact. There was also another issue fixed with the volume name and the artifact name which the previous version of the code expected them to match.

[These](https://sandbox-east.spinnaker.homedepot.com/#/applications/smoketests/executions?pipeline=gke-versioned-volume,gke-versioned-volume2) are the 2 somketests in sandbox east to verify the fix

[helm smoketes](https://np-east.spinnaker.homedepot.com/#/applications/smoketests/executions?pipeline=gke-helm3) in np east where this version of the code was deployed as well to verify the fix

### CODE REVIEW

File: pkg/http/core/kubernetes/deploy.go [hunk 1]
```diff
@@ -134,18 +133,19 @@ func Deploy(c *gin.Context, dm DeployManifestRequest) {
 			return
 		}
 
-		err = kc.VersionVolumes(u, dm.RequiredArtifacts)
+		namespace := u.GetNamespace()
+		err = kc.VersionVolumes(u, namespace, application, client)
+
 		if err != nil {
```
Comment (billiford) [inline pkg/http/core/kubernetes/deploy.go:139]: Let's put defined errors next to their error checks.
```go
		namespace := u.GetNamespace()

		err = kc.VersionVolumes(u, namespace, application, client)
		if err != nil {
```

File: pkg/http/core/kubernetes/deploy.go [hunk 2]
```diff
@@ -134,18 +133,19 @@ func Deploy(c *gin.Context, dm DeployManifestRequest) {
...
 		if kc.IsVersioned(u) {
-			kind := strings.ToLower(u.GetKind())
-			namespace := u.GetNamespace()
 			labelSelector := metav1.LabelSelector{
 				MatchLabels: map[string]string{
-					kubernetes.LabelKubernetesName: application,
+					kubernetes.LabelKubernetesName:      application,
+					kubernetes.LabelKubernetesManagedBy: kubernetes.Spinnaker,
 				},
 				MatchExpressions: []metav1.LabelSelectorRequirement{
```
Comment (billiford) [inline pkg/http/core/kubernetes/deploy.go:150]: Can we move all the logic inside this if statement to a helper function that takes in whatever arguments it needs?
```go
if kc.IsVersioned(u) {
  err = version(u, ...) // Assuming that argument length doesn't exceed three.
  if err != nil {
    // Handle error.
  }
}

File: pkg/http/core/kubernetes/deploy.go [hunk 3]
```diff
@@ -155,12 +155,20 @@ func Deploy(c *gin.Context, dm DeployManifestRequest) {
...
 
+			ls, err := metav1.LabelSelectorAsSelector(&labelSelector)
+			if err != nil {
+				clouddriver.Error(c, http.StatusInternalServerError, err)
+				return
+			}
+
 			lo := metav1.ListOptions{
-				LabelSelector:  labels.Set(labelSelector.MatchLabels).String(),
+				LabelSelector:  ls.String(),
```
Comment (billiford) [inline pkg/http/core/kubernetes/deploy.go:165]: What is the output of this label selector string? Have we verified that it is only pulling down resources that have a version annotation?

File: pkg/kubernetes/version.go [hunk 4]
```diff
@@ -45,6 +49,7 @@ func (c *controller) GetCurrentVersion(ul *unstructured.UnstructuredList, kind,
 	}
 
 	results := manifestFilter.FilterOnClusterAnnotation(cluster)
+
```
Comment (billiford) [inline pkg/kubernetes/version.go:52]: Remove this newline.

File: pkg/http/core/kubernetes/deploy.go [hunk 5]
```diff
@@ -155,12 +155,20 @@ func Deploy(c *gin.Context, dm DeployManifestRequest) {
...
 				TimeoutSeconds: &listTimeout,
 			}
 
+			kind := strings.ToLower(u.GetKind())
 			results, err := client.ListResourcesByKindAndNamespace(kind, namespace, lo)
+
 			if err != nil {
```
Comment (billiford) [inline pkg/http/core/kubernetes/deploy.go:172]: Move the error declaration next to the if statement.

File: pkg/kubernetes/version.go [hunk 6]
```diff
@@ -109,58 +114,101 @@ func (c *controller) IncrementVersion(currentVersion string) SpinnakerVersion {
...
+			volume.Secret.SecretName = currentVersion
 		}
 	}
+
+	return nil
+}
+
+func (c *controller) GetVolumeVersion(name, kind, namespace, application string, kubeClient Client) (string, error) {
+	//get resourcces
```
Comment (billiford) [inline pkg/kubernetes/version.go:181]: Remove this comment.

File: pkg/kubernetes/version.go [hunk 7]
```diff
@@ -109,58 +114,101 @@ func (c *controller) IncrementVersion(currentVersion string) SpinnakerVersion {
...
+		MatchExpressions: []metav1.LabelSelectorRequirement{
+			{
+				Key:      LabelSpinnakerMonikerSequence,
+				Operator: metav1.LabelSelectorOpExists,
+			},
+		},
+	}
+
+	lo := metav1.ListOptions{
+		LabelSelector:  labels.Set(labelSelector.MatchLabels).String(),
```
Comment (billiford) [inline pkg/kubernetes/version.go:196]: This is the incorrect label selector.

**Review by billiford — CHANGES_REQUESTED** [review]
Most of what I commented on is just formatting, but there's a few things I don't like.

1) the new functions declared in the controller interface take in 4 or 5 arguments, which is a lot - it's best to try to not go over three arguments if possible
2) we're passing the kube client out of the `http/core/kubernetes` package which feels weird - we do not do this anywhere else in the code (that I know of)
3) I don't see any tests (maybe I'm blind)

These things lead me to believe we should try to implement this logic *inside* the `http/core/kubernetes/deploy.go` file using helper functions and adding tests. There might be some reasoning that I'm missing, but overall the function declarations and logic look more complicated than they should be.

In short, I'd like it if we could avoid passing the kubernetes client as an argument to the controller and keeping whatever new functions we declare in the controller to have 3 or fewer arguments.

File: pkg/http/core/kubernetes/deploy.go (hunk 1 above)
Comment (yasmin-abdullah) [reply pkg/http/core/kubernetes/deploy.go:139]: done

File: pkg/http/core/kubernetes/deploy.go (hunk 3 above)
Comment (yasmin-abdullah) [reply pkg/http/core/kubernetes/deploy.go:165]: No, I can add a test for that

File: pkg/kubernetes/version.go (hunk 4 above)
Comment (yasmin-abdullah) [reply pkg/kubernetes/version.go]: done

File: pkg/http/core/kubernetes/deploy.go (hunk 5 above)
Comment (yasmin-abdullah) [reply pkg/http/core/kubernetes/deploy.go:172]: done

File: pkg/kubernetes/version.go (hunk 6 above)
Comment (yasmin-abdullah) [reply pkg/kubernetes/version.go]: done

File: pkg/kubernetes/version.go (hunk 7 above)
Comment (yasmin-abdullah) [reply pkg/kubernetes/version.go:194]: fixed, but not tested yet

**Comment by yasmin-abdullah** [comment]
I opened a different pull request because there were some design changes https://github.com/homedepot/go-clouddriver/pull/27
